### ISSUE TITLE
Inconsistent Flicker-free Dimming Behavior

### ISSUE THREAD

**Issue by Velgus** [seerge/g-helper#2543]
### Rules

- [X] I made myself familiar with the <a href='https://github.com/seerge/g-helper?tab=readme-ov-file'>Readme</a>, <a href='https://github.com/seerge/g-helper/wiki/FAQ'>FAQ</a> and <a href='https://github.com/seerge/g-helper/wiki/Troubleshooting'>Troubleshooting</a>.
- [X] I understand that, if insufficient information or no app logs will be provided, my issue will be closed without an answer.

### What's wrong?

Visual modes seem to have inconsistent application combined with Flicker-free Dimming. For example, if switching from one mode to another with Flicker-free Dimming at 50%, it will be substantially brighter, but then if you click 50% in Flicker-free dimming, it will be appropriately dimmed.

This also happens when opening the Windows "System > Display" settings menu.

This doesn't seem to apply to all modes, but I found one where I could reproduce it consistently (see "How to reproduce the bug").

### How to reproduce the bug?

1. Set Flicker-free Dimming to 50% on "Gamut: Native".
2. Switch to "Gamut: DCIP3" - notice that it seems overly bright for 50%.
3. Click 50% in Flicker-free Dimming again (exactly where the slider is) - notice it gets dark again.
4. Switch back to "Gamut: Native" - notice that this switch seems to be appropriately dark, and doesn't change if you click 50% in Flicker-free Dimming.

### Logs

[log.txt](https://github.com/seerge/g-helper/files/15206526/log.txt)

### Device and Model

Asus Zephyrus G14 GA403

### Additional information.

I looked for the "Automatically manage color for apps" setting, in case it was interfering, but that setting doesn't exist on my model.

### Armoury Crate

Uninstalled

### Asus Services

None

### Version

0.170

### OS

Windows 11 23H2

**Comment by seerge** [comment]
@Velgus hello, G-Helper just runs ```AsusSplendid.exe``` with appropriate params to set both dimming and gamuts, same as AC does.

How does this work if you would repeat same steps using AC ?

**Comment by seerge** [comment]
@Velgus hello, any updates here? Did you try same thing in AC ?

**Comment by Velgus** [comment]
Sorry, I'll get back to you soon.

**Comment by Velgus** [comment]
Okay, so after testing, the main issue is more to do with the settings being confusing, then a bug with switching the flicker-free dimming. The issue exists in AC as well, but if you want, you could make a slight adjustment to make it a bit clearer for G-Helper users.

Basically, flicker-free dimming only works with with "Gamut: Native" mode. If you are in any other mode, adjusting flicker-free dimming switches to "Gamut: Native" mode and adjusts the dimming. However, the UI doesn't update (in G-Helper or AC), so it will still say it is in another mode (eg. "Gamut: DCIP3" or whatever), even though it is not actually using that mode.

It could be clarified by simply automatically switching to the "Gamut: Native" mode in the drop-down when you change flicker-free dimming.

Additionally, I'm not totally sure why, but entering the Windows Display settings exhibits weird behavior as mentioned before. As far as I can tell what seems to happen is, if you are in "Gamut: Native" mode, it switches you to the last non-"Gamut: Native" mode you were in (though again, no UI update in G-Helper or AC).

**Comment by seerge** [comment]
@Velgus hello, thanks for checking

> Basically, flicker-free dimming only works with with "Gamut: Native" mode. If you are in any other mode, adjusting flicker-free dimming switches to "Gamut: Native" mode and adjusts the dimming. 

That's an interesting find. I don't have G14 2024 anymore to double check tho

>It could be clarified by simply automatically switching to the "Gamut: Native" mode in the drop-down when you change flicker-free dimming.

I think this can be an option. I'm not a fan of hiding dimming from the UI. As it's also accessible via hotkeys. It's easier to update ``Gamut`` dropdown.

But I would like to double check if it indeed resets gamut to default (de-facto) from someone else with same device

**Comment by Velgus** [comment]
>  I'm not a fan of hiding dimming from the UI. As it's also accessible via hotkeys. 

Do you know which hotkeys it's supposed to be? The FN+F7/F8 adjusts the regular display brightness, not Flicker-free Dimming.

**Comment by seerge** [comment]
@Velgus Shift + FN + F7/F8

**Comment by Velgus** [comment]
Ah, found it, thanks. It's Ctrl, not Shift though. Shift + FN + F7/F8 is Slash Lighting Brightness.

**Comment by seerge** [comment]
@Velgus right :) sorry. I have also added this to the readme.

**Comment by lgarlati** [comment]
G16 2024 here, and yeah. but I don't know if it's a bug or just the calibration affecting brightness. 
Native = DisplayP3 = sRGB > DCIP3

**Comment by seerge** [comment]
@lgarlati ok, so dimming does not reset (actual) gamut on your device ?

**Comment by lgarlati** [comment]
@seerge This is even weirder than I thought. I ran the following test with each other three color modes, but DCIP3 was the most obvious:
Set to any mode other than native, dim 10% via hotkey, brighten 10% via hotkey, switch to native to see difference. 

Some notes at each step:
- At 100% brightness, all the gamut modes work as expected.
- upon dimming, the gamut visually changes, a lot. This new gamut resembles native.
- upon brightening, there is no gamut change compared to dimmer, just a brighter screen.
- Upon switching back to native, there *IS* a change, albeit slight. This is most obvious from DCIP3, but present from the others as well. 
- Lastly, once back at 100% switching between modes works again as intended. 

Not sure what to make of this, but it seems that dimming is messing with the gamuts themselves.

**Comment by seerge** [comment]
@Velgus @lgarlati great, thanks for explanation of your observations :)

Dimming is definitely "messing" with gamuts. And it is supposed to be so.

What dimming effectively does - it compresses dynamic range of the actual image, by just reducing it's brightness.
It's pretty much if you would take a screenshot and then adjust it's brightness in photoshop to make darker.

Given that it seem to return back to original state when you return brightness, I think it doesn't make sense to change anything in app UI. And just keep currently selected gamut as it was before (also to be consistent with how AC shows this)

**Comment by lgarlati** [comment]
> Given that it seem to return back to original state when you return brightness, I think it doesn't make sense to change anything in app UI.
I must have made my previous post unclear. It does NOT recover once you return to 100%. It remains as a weird near-native gamut. You need to manually reselect the desired gamut for it to take effect.

**Comment by seerge** [comment]
@Velgus @lgarlati 

Ok, then I can reset Gamut dropdown to default value every time you adjust dimming


Try this build [GHelper.zip](https://github.com/user-attachments/files/15613185/GHelper.zip)

**Comment by lgarlati** [comment]
@seerge 

> Ok, then I can reset Gamut dropdown to default value every time you adjust dimming

Can confirm that it does reset, only problem is it resets to native, not the previous choice. Also, I think resetting every time the dimming is changed it unnecessary and adds even more latency while sliding the slider. May be better to make it reset whenever the dimming returns to 100% (not dimmed)

**Comment by seerge** [comment]
@lgarlati 

>Can confirm that it does reset, only problem is it resets to native, not the previous choice.

It resets UI to Native intentionally. Cause as I have understood you both correctly - any Dimming levels reset gamut back to Native (and don't recover after) ? Change is purely cosmetic. I can't change actual dimming behavior anyhow.

>Also, I think resetting every time the dimming is changed it unnecessary and adds even more latency while sliding the slider. May be better to make it reset whenever the dimming returns to 100% (not dimmed)

It resets dropdown only when it's not native currently. So it should happen only once :)

**Comment by lgarlati** [comment]
@seerge 

> any Dimming levels reset gamut back to Native (and don't recover after) ? 

Sorry I wasn't clearer. Once dimmed, the gamuts change and resemble native. But they are NOT the same. There is a distinct difference, effectively an entirely new set of gamuts.

**Comment by lgarlati** [comment]
@seerge Understood. I just meant keep what you added in the zip above, just set it to the last gamut instead of native.

**Comment by lgarlati** [comment]
> But that would be a behavior of current (regular) release. I.e. no changes at all ?

Yes, but reapplying the same gamut once back to 100% fixes the issue. 

If you can point me to where in the codebase this is implemented, I can try playing around with it to see what works.

**Comment by seerge** [comment]
@lgarlati yeah, but whole discussion here is only about adjusting UI to reflect actual state of the things with dimming / gamuts under the hood (they are controlled by Asus Utility not by G-Helper).

So if adjusting Dimming **does not** reset custom gamuts (like DCI-P3 or so) to Native. I think it's better to keep old gamut selected in UI.

I just thought that Dimming always resets any Gamut to Native (which is not the case).

So the changes i did in the build here are probably not needed.

Anyway, this is where everything is handled in a code
https://github.com/seerge/g-helper/blob/main/app/Display/VisualControl.cs

**Comment by lgarlati** [comment]
@seerge 

I'll take a look after work and see what I can do. I'll report back either tonight or tomorrow.

**Comment by lgarlati** [comment]
@seerge 

First, it turns out that my previous assumption that there was a "second set" of gamuts while dimmed is false. By reapplying a gamut, dimmed or otherwise, the correct gamut is shown, dimmed. As such, I found a janky fix. Simply calling "setgamut" again after a delay fixes the problem. While my delayed reapply works, the true cause I'm unsure of. I suspect somewhere in the brightness setting pipeline another erroneous call is made to set gamut which is resetting away from user preference. 

Here is the changes I made. First adding this function:

```

        private static void SetGamutDelayed(int mode, int delay)
        {
            static void GamutDelayElapsed(object? sender, System.Timers.ElapsedEventArgs e, int mode)
            {
                Logger.WriteLine("Gamut Delay Timer Elapsed, setting gamut to "+mode);
                GamutDelayTimer.Stop();
                GamutDelayTimer.Dispose();
                GamutDelayTimer = default!;
                SetGamut(mode);
            }
            if (GamutDelayTimer is null)
            {
                Logger.WriteLine("Delayed Gamut set to"+mode+" with delay "+delay+"ms");
                GamutDelayTimer = new System.Timers.Timer(delay);
                GamutDelayTimer.Elapsed += (sender, e) => GamutDelayElapsed(sender, e, mode);
                GamutDelayTimer.AutoReset = false;
                GamutDelayTimer.Start();
            }
            else
            {
                Logger.WriteLine("Gamut Delay Timer already running");
            }
        }
```
with a corresponding timer at the top. Then adding a callback to it on the last line of `SetBrightness`
```
SetGamutDelayed(AppConfig.Get("gamut", -1), 1000);
```
I used a delay of 1 second as it shows the switch back fixing the issue, but a smaller value, say 500, likely also works.

**Comment by seerge** [comment]
@lgarlati  ok, i see. 

How does it work when you do dimming in AC ? (i was asking @Velgus here already) But I wonder if it even tries to handle this ?

**Comment by Velgus** [comment]
Sorry, I lost track of this thread (for some reason I wasn't getting notified by mentions and such), but Igarlati's observations are likely more accurate than my own. That said, when I tested with AC it exhibited the exact same behaviors that G-Helper currently does. So basically AC "doesn't" handle the issue as far as I could tell, and any handling added here would be an improvement over the way AC works.

**Comment by seerge** [comment]
@Velgus you can check build i posted above. It resets UI (i.e. gamma dropdown) to Native every time you change Dimming.

**Comment by Velgus** [comment]
The custom build still exhibits the same strange behavior that Igarlati and I both experienced. While it switches back to Native, it still has a weird unknown Gamut setting, and it only corrects when you click a value in the slider again (even the same value). 

As such, yeah, I think my initial observation about it defaulting to Native was incorrect and Igarlati's solution would be better.

**Comment by lgarlati** [comment]
While my solution definitely _works_ it also is very, very janky and not robust.

The correct solution would be to find wherever the code that's incorrectly resetting the gamut is, and place the correct setting code right after it. However, that sounds like a lot of work. As for what delay to use, I'd say somewhere between 500 and 1000ms.

**Comment by seerge** [comment]
@lgarlati 

> The correct solution would be to find wherever the code that's incorrectly resetting the gamut is, and place the correct setting code right after it. However, that sounds like a lot of work.

There is no access to the "code" responsible for that, cause as you see it's just AsusSplendid.exe that does it inside :)

Honestly looking at all this, I'm not a big fan of manually calling AsusSplendid and setting gamut every single time Dimming is changed. Especially considering that AC doesn't do that neither.

**Comment by lgarlati** [comment]
@seerge 

> I'm not a big fan of manually calling AsusSplendid and setting gamut every single time Dimming is changed. Especially considering that AC doesn't do that neither.

I personally lean in favor of having a delayed re-application of the gamut. It's really annoying to have to switch it back twice every day when I use dimming at night. But, I would make it opt-in via "extra" menu, and would make the 1 second timer reset every time the brightness is adjusted so it doesn't get reset multiple times ie 
```
if(timer already running){
    timer.stop()
    timer.start()
}
```
instead of the 
```
else
{
    Logger.WriteLine("Gamut Delay Timer already running");
}
```

Regardless of what you decide, I'm definitely keeping those few lines of code in my version of the app XD

**Comment by NineBallo** [comment]
I've had this happen when opening the bottom right submenu in windows (the one with brightness, volume, bluetooth, etc...) it will just reset the display's flicker free dimming to 100% until manually set back to the correct/displayed value in g-helper. However ghelper will still display the old value. I presume something is setting the value in windows, and g-helper cannot read state only set it; hence the lack of update in the program, and the value fixing once the slider is touched. Is there any way I can help debug this? 

Zephyrus GA605

**Comment by seerge** [comment]
@NineBallo hello, what you explain is not the same as being discussed in this topic.

Anyway - what "flicker free dimming" does, is just lowers display gamma i.e. makes screen darker in a software way. It's handled not by G-Helper, but by `AsusSplendid` utility (part of Asus Control Interface package). I can imagine windows just resets gamma when you interact with bottom right submenu by some reason. Do you use asus version for AMD drivers or ones from official AMD website ? Try to update them

**Comment by NineBallo** [comment]
> @NineBallo hello, what you explain is not the same as being discussed in this topic.
> 
> Anyway - what "flicker free dimming" does, is just lowers display gamma i.e. makes screen darker in a software way. It's handled not by G-Helper, but by `AsusSplendid` utility (part of Asus Control Interface package). I can imagine windows just resets gamma when you interact with bottom right submenu by some reason. Do you use asus version for AMD drivers or ones from official AMD website ? Try to update them

Ah well apologies then, I'm using the amd drivers just because they are generally more up to date. Should I use the asus ones instead? I wonder if it's a hdr thing, it might be windows enforcing the gamma due to hdr; i know brightness is often enforced so maybe? I'll see if I can reproduce it with hdr off and start a new issue so I'm not hijacking this one if it persists. Thanks for the help!

**Comment by thatonehhdev** [comment]
> @NineBallo hello, what you explain is not the same as being discussed in this topic.
> 
> Anyway - what "flicker free dimming" does, is just lowers display gamma i.e. makes screen darker in a software way. It's handled not by G-Helper, but by `AsusSplendid` utility (part of Asus Control Interface package). I can imagine windows just resets gamma when you interact with bottom right submenu by some reason. Do you use asus version for AMD drivers or ones from official AMD website ? Try to update them

yes, windows does reset screen stuff when you access the bottom right menus. on my old laptop, the screen calibration wouldn't apply until I clicked something in that menu.

**Comment by seerge** [comment]
@Velgus @lgarlati @NineBallo @ouya-woelders 

Hello, I currently use G14 2025 (with latest Asus System Control Interface as it includes `Asus Splendid` app responsible for dimming / gamuts). And I actually can not reproduce original scenario. I.e. if I set dimming to some value, it persists even after changing Gamut or Visal Mode. So looks like Asus has fixed it.

Can you check on your devices (preferably update ASCI to the latest too) if this is the case ?

Thanks

**Comment by NineBallo** [comment]
@seerge 
G16 2025 GA605 

The issues I faced were from an AMD driver bug if I remember correctly. Had a ton of them for a few months and was driving me mad. Was eventually fixed and can now use flicker free without any of the issues I originally faced.

Regarding this issue: yep can confirm it seems to be fixed. The dimming value persists when changing between native, srgb, dcip3 at least on my machine.

**Comment by ouya-woelders** [comment]
im pretty sure this is a windows bug, i happens to me on my other laptop too (xiaomi, not asus)

**Comment by seerge** [comment]
Then I assume it's "resolved" by itself (or rather by asus update). I'm glad it doesn't require any workarounds.

Feel free to post an update if something changes.

Thanks
